### ISSUE TITLE
would be nice in future: export html overview of all installed / available stickerpacks (stickerpacks.html)

### ISSUE THREAD

**Issue by Meteor0id** [bepaald/signalbackup-tools#117]
In an effort to export anything a backup includes, it might be a useful feature to some people to see what stickers they have in their signal backup, meaning an overview of all sticker packs and each sticker in that pack. If possible also a list of which chats the pack is associated with / used in.

This would be some advanced feature for later concern obviously.

**Comment by bepaald** [comment]
This has been a long time for sure, but I just implemented this. Undocumented at the moment, to see if there is any feedback to deal with (if you're still around and interested of course).

Adding `--stickerpacks` to `--exporthtml` will generate a page of installed stickerpacks, accessible from the index page. If both stickerpacks and calllogs are requested the two links should move into a collapsible menu on the index page. Any feedback is welcome, though I'm not particularly looking forward to redesigning the entire layout...

(as a side note, there is now also `--importstickers` as a modifier to `--importfromdesktop`, to import installed stickerpacks from a Desktop database. untested!)

**Comment by Meteor0id** [comment]
Really nice to see this feature added! Makes the tool very complete. But first a few bugs

![menu](https://github.com/bepaald/signalbackup-tools/assets/34976212/6afe303d-5937-4201-815e-32f87e6dc4df)
![stickerpacks no render](https://github.com/bepaald/signalbackup-tools/assets/34976212/bdb92224-f63b-4b38-98a0-e473a9d50078)

**Comment by bepaald** [comment]
First one is fixed. It was a stupid variable mixup. Apparently I don't actually open the final output when I finally paste the new code into the source file, or I would have caught that.

Second one is strange. Do all stickers look like that, or just this one pack? Is it a pack I could maybe install? I assume these are imported from Desktop? Did the import print any errors while importing stickerpacks? If you find them (spread over sub-directories of `HOME/AppData/Roaming/Signal( Beta)/stickers.noindex/`) do they look ok? Not sure if Windows has some tool to get file properties, but those might be interesting, for example on my machine:
```
$ file ~/.config/Signal/stickers.noindex/0c/*
~/.config/Signal/stickers.noindex/0c/0cafe3a4be9177a521327603e931fbd30fa0be479120527ef197694c1632cb29: RIFF (little-endian) data, Web/P image
~/.config/Signal/stickers.noindex/0c/0cf3a5499e758050e0fe7f223122b72a83c4e5592e45a13484f72145e79a9562: PNG image data, 512 x 512, 8-bit colormap, non-interlaced
```

I have a suspicion something is wrong on the Desktop import side of things here.

Thanks!

**Comment by Meteor0id** [comment]
All of the stickers of any pack look that way and as you can see a random sticker there did render some odd faulty output but that's probably meaningless.

I've checked the Desktop files and am able to simply open and view them using irfanview (an image viewer). The files are fine.


For the record, one day before you published the new sticker export function I cleaned up my Signal Desktop, deleting a whole lot of conversations which I had left there for testing this tool :p. But there appears to be plenty left to test with.

**Comment by Meteor0id** [comment]
Printing the stickers in my case shows one sticker (well I can't see the sticker at this time but I can see the layout) on each line. I haven't debugged it much but it may have to do with some of my stickerpacks having a whole lot of emojis to them. Here you have one of those stickerpacks to test with: https://signalstickers.com/pack/ee7c8f7152bb5203d88c8d2cfb9d71ba
It may look better if those emojis line-wrap.

**Comment by bepaald** [comment]
I fired up a VM and was able to reproduce your second issue, seems to be a Windows thing. Now that I have it myself I hope I can find the problem and fix it.

>Printing the stickers in my case shows one sticker (well I can't see the sticker at this time but I can see the layout) on each line. I haven't debugged it much but it may have to do with some of my stickerpacks having a whole lot of emojis to them. 

I'll have a look at this later as well. I don't know if it's the multiple emoji's, but that's interesting anyway, I didn't even know sticker could have more than one.

**Comment by bepaald** [comment]
Ok, it was an old but forgotten Windows quirk: I needed to explicitly open the output file in binary mode.

I've done some css tweaks to make the cells grow taller (instead of wider) when there are a lot of emoji. I think it's an improvement, but I'm not sure if there is a nicer way to deal with those cases.

I've looked at the printing thing, but haven't figured it out yet. I thought it might have something to do with the `@media screen` stuff I'm doing, but removing those rules has no effect.

**Comment by bepaald** [comment]
I may have fixed it by removing `display: block;` and `margin: 0 auto;` from the sticker-list class in `@media print`. But I don't know if that has other side effects (I don't know why these two things were there, I copied them from one of the other html files).

I did notice in the print preview, some stickers appear to be broken at page boundaries (even though there is already a `break-inside: avoid;` on the sticker-list-item class. Not sure if that's just a bad preview or something else to fix.

**Comment by Meteor0id** [comment]
`I did notice in the print preview, some stickers appear to be broken at page boundaries (even though there is already a break-inside: avoid; on the sticker-list-item class. Not sure if that's just a bad preview or something else to fix.`

yea that one I noticed too and came here to report it. I haven't figured out why it isn't working, even applying position:relative; to parent and child didn't work. Not sure what goes on here but it's pretty useless to print these packs wit hthe stickers cut in half.

furthermore if you print while zoomed on an image the zoomed image is printed enlarged, not sure if this is avoidable since I think you are using javascript here? I didn't check, but you could try @media print { something to disable zoomed images even ones currently active }

Overall the sticker export looks good, just not when printing.

**Comment by Meteor0id** [comment]
Also when I refresh the page while an image is zoomed, it does not unzoom.

**Comment by bepaald** [comment]
From a few quick searches it seems it may have to do with its (or its parents) `display: flex`, or `flex-wrap` property. Though in all cases it's probably a bug in the browser, and not against css spec.

I'm not at home this weekend, but I'll try to work around it when I'm back. Not sure if I can though.

The image zooms use no javascript but a hidden checkbox element (a fairly common method of doing these sorts of thing in pure css, the collapsible menu works the same way). It should be quite simple to undo the zoom in print mode. As for page refresh, I guess Firefox just remembers the values of input elements on refresh. I probably can't do much about that, but I don't consider that one a big deal.

**Comment by Meteor0id** [comment]
I dont know why I assumed you were using javascript i wasn't even looking.

You can fix the refresh isshe by reversibg the way you use checkboxes.

`<Input type="checkbox" checked="checked" />`
And in css you zoom when a box is unchecked. On any reload all boxes get checked again because that is hardcoded.

**Comment by bepaald** [comment]
>On any reload all boxes get checked again because that is hardcoded.

Did you test this? I thought it was a clever solution, but it doesn't seem to work for me. Firefox has a 'feature' that it remembers the user supplied values in input fields and forms on soft refresh (not on hard refresh, CTRL F5). Whether the value starts out checked or unchecked doesn't seem to matter in this case. Firefox also supplies a workaround with the `autocomplete=off` attribute (which per spec is not valid on `input`s with `type=checkbox`, but Firefox allows it specifically to work around this 'feature'). So I implemented that (though to still get through validation I added a `form`-container and set the property on that.

I also added some css to the `@media print` section to undo any zoomed in stickers in print mode.

That leaves the page-breaking. I won't be looking at that today, hopefully tomorrow or the day after, but I still don't know how (or if) I can fix it.

**Comment by Meteor0id** [comment]
I've rewritten the @media print section.
- fit 4 stickers side by side
- avoid breaking stickers across pages
- keep emojis centered despite not using flex

```
      @media print {
        .sticker-list-header {
          padding: 0;
        }

       .sticker-list-item {
          break-inside: avoid;
		  margin: 3px;
        }

        .sticker-list {
          border-radius: 0;
		  padding-left: 0;
		  padding-right: 0;
        }

        .sticker img,
        .sticker input[type=checkbox]:checked ~ label > img { 
          z-index: 1;
          cursor: default;
          background-color: #00000000;
          border-radius: 0px;
          border: none;
          padding: 0px;
          transition: none;
          transform: none;
        }

        #menu {
          display: none;
        }

        #theme {
          display: none;
        }
		.sticker-list-item, .sticker label {
		display: unset;
		}
		.sticker-list-item .footer {
		display: table;
		margin-left: auto;
		margin-right: auto;
		}
      }
```

**Comment by bepaald** [comment]
Thanks for your efforts, I was going to try and convert some elements to `block` elements, but you were one step ahead. I opted to apply it mostly to the main css, since it was visually equivalent anyway. I did need to make some adjustments as some stickers would not center correctly or would grow outside their boxes, though those issues might have come up with my last push.

I also used `display: block` instead of `table` on the footer class, but maybe I'm missing why you chose `table` specifically for that one?

Lastly I applied your suggested padding/margin and added some styling to change the width of the stickerlist with width of the print.

I do notice, that `sticker-list-item`s still actually break, but now only between the sticker and the footer. That still seems like an implementation bug, though not as bad as in the middle of an `img`. Is that acceptable?

Are you seeing any more issues or problems?

Thanks!

**Comment by Meteor0id** [comment]
Looks perfect to me, would say this feature is done.
Display table was just my quick way of centering the footer you did it another way and it looks all good.

**Comment by Meteor0id** [comment]
Actually, is there anything we can say about whether the pack was installed or whether it is only available but not installed yet? I guess you're only including installed packs atm?

**Comment by bepaald** [comment]
>Actually, is there anything we can say about whether the pack was installed or whether it is only available but not installed yet? I guess you're only including installed packs atm?

Correct.

I gave it a go. But (I'm guessing this is your next request) they're not imported from desktop yet. I'll get on that, hopefully sometime in the next few days, but maybe over the weekend depending on how busy work gets.

**Comment by bepaald** [comment]
I think importing known stickerpacks is fro desktop is now implemented. It's somewhat tricky to test due to a bug/change in behavior of Signal Desktop (I've opened an issue) where new transactions are not necessarily written to the database when the app closes (they are left in the temporary file). This causes the database to appear slightly out of date, so I can't quickly send a message (or sticker) to Desktop and test the results.

Anyway, I do think this is done. Is everything still looking good since the last changes (outputting 'available' packs)? If so, I'll update the readme and `--help` and close this one...
